Wasm/Engine.lean

```diff
@@ -180,139 +165,152 @@ def findLocalByName? (ls : List (Option String × Option StackEntry))
+def σmap : (List StackEntry → List StackEntry) → EngineM PUnit :=
+  fun f => do set $ f (←get)
+def bite : EngineM StackEntry := do match (←get) with
+  | [] => throwEE .not_enough_stuff_on_stack
+  | s :: rest => set rest; pure s
+
 mutual
-  /- TODO: Support multi-output functions. -/
+
   partial def getSO (locals : List (Option String × Option StackEntry))
-                    (stack : List StackEntry)
-                    : Get'
-                    → Except EngineErrors (ContinuationStack × StackEntry)
-    | .from_stack => match stack with
-      | [] => .error .not_enough_stuff_on_stack
-      | s :: rest => .ok ((rest, .none), s)
-    | .from_operation o => do
-      -- Some instructions do not produce a value/do not change stack.
-      let (stack', cont?) ← runOp locals stack o
-      match stack' with
-      | [] =>
-        if cont?.isSome
-          then throw .other -- TODO: this is incorrect! It's possible to `br` to
-                            -- a resultless block, in which case the empty stack
-                            -- should be returned with the continuation, but rn
-                            -- it's not possible. Continuations should ideally
-                            -- be passed in a monadic way, not this jackally.
-          else throw .not_enough_stuff_on_stack
-      | s :: rest => pure ((rest, cont?), s)
+                    : Get' → EngineM StackEntry
+    | .from_stack => bite
+    | .from_operation o => do runOp locals o; bite
     -- TODO: names are erased in production. See what do we want to do with this code path.
     | .by_name n => match n.name with
-      | .none => .error .local_with_no_name_given
+      | .none => throwEE .local_with_no_name_given
       | .some name => match findLocalByName? locals name with
-        | .none => .error $ .local_with_given_name_missing name
-        | .some l => .ok ((stack, .none), l)
+        | .none => throwEE $ .local_with_given_name_missing name
+        | .some l => pure l
     | .by_index i => match locals.get? i.index with
-      | .some (_, .some se)  => .ok ((stack, .none), se)
-      | _ => .error $ .local_with_given_id_missing i.index
+      | .some (_, .some se) => pure se
+      | _ => throwEE $ .local_with_given_id_missing i.index
 
-  partial def runWithContinuation
+  partial def computeContinuation
                       (blocktypes : List Type')
                       (locals : List (Option String × Option StackEntry))
-                      (so : ContinuationStack)
                       (ops' : List Operation)
-                      : Except EngineErrors (List StackEntry) := do
+                      : EngineM PUnit := do
     let rec go
-    | (stack, .none), [] => pure stack
-    | (stack, .some cont), _ => go (stack, .none) cont
-    | (stack, .none), op :: ops => do go (←runOp locals stack op) ops
+    | [] => pure ()
+    | op :: ops => do match ←(runOp locals op).run (←get) with
```

Comment:
Syntax so cool my brain froze for a second. Just in case someone's confused, I think this will allow us to match against a tuple where left element is the result of `runOp` and the right is the contains of the state, which happen to be List StackExpression (see above). We then match and jump to execute `cont`, which happens to be of type of whatever is stored on the **outermost** `Except` layer of the monad. In our case, it's `Continuation`, which is just a `List Operation`.

If we look at how `runOp` uses "throw", we see that it is so far used only in one place, here:

```
    | .br li => checkLabel li fun ⟨n, cont⟩ => do
      let (topn, rest) := (←get).splitAt n
      if (stackValues ⟨topn⟩).length = n
        then match skimValues ⟨rest⟩ with
          | .label _ :: bottom =>
            set $ topn ++ bottom
            raiseCont $ if li = ⟨0⟩ then cont else [.br ⟨li.li-1⟩]
          | _ => throwEE .typecheck_failed
        else throwEE .not_enough_stuff_on_stack
```

The interesting bit is `raiseCont`. This is an alias for `throw` which is clearly declares that branching is happening.

I'll leave the reader of my review with this code, this is another place where the payoff for these changes is happening. :)